File: index1.py
#!/usr/bin/env python
#!coding:utf-8

#方案一：8个小灯逐个闪烁（跑马灯程序），用户手离超声波传感器遮挡进入一定范围小灯全亮(10厘米)
#Author: Jiahui Tang
#Time: 2016-11-30

#import the necessary packages
import RPi.GPIO as GPIO
import time

#define the led
led1 = 29
led2 = 31
led3 = 33
led4 = 35
led5 = 37
led6 = 36
led7 = 38
led8 = 40

#the array of the led
led = [led1, led2, led3, led4, led5, led6, led7, led8]

#the distance detection, by the HC-SR04
def checkdist():
	GPIO.output(11, GPIO.HIGH)
	time.sleep(0.000015)
	GPIO.output(11, GPIO.LOW)
	while not GPIO.input(12):
		pass
	t1 = time.time()
	while GPIO.input(12):
		pass
	t2 = time.time()
	
	return (t2-t1) * 340 / 2

#set the GPIO's mode
GPIO.setmode(GPIO.BOARD)

#initialize the led
for i in led:
	GPIO.setup(i, GPIO.OUT, initial = GPIO.LOW)

	
#the i/o(Trig, Echo) of the HC-SR04
GPIO.setup(11, GPIO.OUT, initial = GPIO.LOW)
GPIO.setup(12, GPIO.IN)

#define a temp value
temp = 0

#ready
time.sleep(2)

#main
try:
	#loop
	while True:
		if checkdist() < 0.2:
			for i in led:
				GPIO.output(i, GPIO.HIGH)
		else:
			for i in led:
				GPIO.output(i, GPIO.LOW)
			GPIO.output(led[temp], GPIO.HIGH)
			if temp == 7:
				temp = 0
			else:
				temp += 1
	
		#time delay
		time.sleep(0.5)
		
		#print the distance
		print '%0.2f m' %checkdist()

#security measures
except KeyboardInterrupt:
	GPIO.cleanup()

File: index1.c
//方案一：8个小灯逐个闪烁（跑马灯程序），用户手离超声波传感器遮挡进入一定范围小灯全亮（20厘米）
//Author: Jiahui Tang
//Time: 2016-12-01

//import the necessary packages
#include <wiringPi.h>
#include <stdio.h>
#include <sys/time.h>

//define the HC_SR04/led
#define Trig	0
#define Echo	1
#define led1    21
#define led2    22
#define led3    23
#define led4    24
#define led5    25
#define led6    27
#define led7    28
#define led8    29

//init
void ultraInit(void)
{
	pinMode(Echo, INPUT);
	pinMode(Trig, OUTPUT);

	pinMode(led1, OUTPUT);
	pinMode(led2, OUTPUT);
	pinMode(led3, OUTPUT);
	pinMode(led4, OUTPUT);
	pinMode(led5, OUTPUT);
	pinMode(led6, OUTPUT);
	pinMode(led7, OUTPUT);
	pinMode(led8, OUTPUT);
}

//the distance detection, by the HC-SR04
float disMeasure(void)
{
	struct timeval tv1;
	struct timeval tv2;
	long start, stop;
	float dis;

	digitalWrite(Trig, LOW);
	delayMicroseconds(2);

	digitalWrite(Trig, HIGH);
	delayMicroseconds(10);	  
	digitalWrite(Trig, LOW);
										
	while(!(digitalRead(Echo) == 1));
	//get the time of now
	gettimeofday(&tv1, NULL);		   

	while(!(digitalRead(Echo) == 0));
	//get the time of now
	gettimeofday(&tv2, NULL);		   

	//the time(us)
	start = tv1.tv_sec * 1000000 + tv1.tv_usec;   
	stop  = tv2.tv_sec * 1000000 + tv2.tv_usec;

	//the distance
	dis = (float)(stop - start) / 1000000 * 34000 / 2;  

	return dis;
}

int main(int argc, char* argv[])
{
	//the array of the led
	int led[8] = {21, 22, 23, 24, 25, 27, 28, 29};
	//define a temp value
	int temp = 0;
	//the distance
	float dis;

	if(wiringPiSetup() == -1)
	{ //when initialize wiring failed,print messageto screen
		printf("setup wiringPi failed !");
		return 1; 
	}

	ultraInit();
									
	digitalWrite(21, LOW);
	digitalWrite(22, LOW);
	digitalWrite(23, LOW);
	digitalWrite(24, LOW);
	digitalWrite(25, LOW);
	digitalWrite(27, LOW);
	digitalWrite(28, LOW);
	digitalWrite(29, LOW);
	
	//loop
	for(;;)
	{
		dis = disMeasure();
		
		if(dis < 20)
		{
			digitalWrite(led1, HIGH);
			digitalWrite(led2, HIGH);
			digitalWrite(led3, HIGH);
			digitalWrite(led4, HIGH);
			digitalWrite(led5, HIGH);
			digitalWrite(led6, HIGH);
			digitalWrite(led7, HIGH);
			digitalWrite(led8, HIGH);
		}
		else
		{
			digitalWrite(led1, LOW);
			digitalWrite(led2, LOW);
			digitalWrite(led3, LOW);
			digitalWrite(led4, LOW);
			digitalWrite(led5, LOW);
			digitalWrite(led6, LOW);
			digitalWrite(led7, LOW);
			digitalWrite(led8, LOW);
		
			digitalWrite(led[temp], HIGH);
			if(temp == 7)
				temp = 0;
			else
				temp += 1;
		}

		//time delay
		delayMicroseconds(1000000);
		
		//print the distance
		printf("%0.2f m\n", &dis);
	}

	return 0;
}
